mlir/lib/Catalyst/Transforms/TBAAPatterns.cpp

```diff
@@ -0,0 +1,120 @@
+        newOp.setTBAATags(ArrayAttr::get(ctx, tag));
+    }
+    else if (isa<FloatType>(baseType)) {
+        if (baseType.isF32()) {
+            tag = tree->getTag("float");
+        }
+        else if (baseType.isF64()) {
+            tag = tree->getTag("double");
+        }
+
+        newOp.setTBAATags(ArrayAttr::get(ctx, tag));
+    }
+    else if (isa<MemRefType>(baseType)) {
+        tag = tree->getTag("any pointer");
+
+        newOp.setTBAATags(ArrayAttr::get(ctx, tag));
+    }
+}
+
+template <typename... Types> bool isAnyOf(Type baseType) { return (isa<Types>(baseType) || ...); }
+
+struct MemrefLoadTBAARewritePattern : public ConvertOpToLLVMPattern<memref::LoadOp> {
+    using ConvertOpToLLVMPattern<memref::LoadOp>::ConvertOpToLLVMPattern;
+
+    template <typename... Args>
+    MemrefLoadTBAARewritePattern(catalyst::TBAATree &tree, Args &&...args)
+        : ConvertOpToLLVMPattern(std::forward<Args>(args)...), tree(&tree){};
+
+    LogicalResult matchAndRewrite(memref::LoadOp loadOp, memref::LoadOpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        auto type = loadOp.getMemRefType();
+        auto baseType = type.getElementType();
+        Value dataPtr = getStridedElementPtr(loadOp.getLoc(), type, adaptor.getMemref(),
+                                             adaptor.getIndices(), rewriter);
+        auto op = rewriter.replaceOpWithNewOp<LLVM::LoadOp>(
+            loadOp, typeConverter->convertType(type.getElementType()), dataPtr, 0, false,
+            loadOp.getNontemporal());
+
+        if (isAnyOf<IndexType, IntegerType, FloatType, MemRefType>(baseType)) {
+            setTag(baseType, tree, loadOp.getContext(), op);
+        }
+        else {
+            return failure();
+        }
+        return success();
+    }
+
+  private:
+    catalyst::TBAATree *tree = nullptr;
+};
+
+struct MemrefStoreTBAARewritePattern : public ConvertOpToLLVMPattern<memref::StoreOp> {
+    using ConvertOpToLLVMPattern<memref::StoreOp>::ConvertOpToLLVMPattern;
+
+    template <typename... Args>
+    MemrefStoreTBAARewritePattern(catalyst::TBAATree &tree, Args &&...args)
+        : ConvertOpToLLVMPattern(std::forward<Args>(args)...), tree(&tree){};
+
+    LogicalResult matchAndRewrite(memref::StoreOp storeOp, memref::StoreOpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        auto type = storeOp.getMemRefType();
+        auto baseType = type.getElementType();
+
+        Value dataPtr = getStridedElementPtr(storeOp.getLoc(), type, adaptor.getMemref(),
+                                             adaptor.getIndices(), rewriter);
+        auto op = rewriter.replaceOpWithNewOp<LLVM::StoreOp>(storeOp, adaptor.getValue(), dataPtr,
+                                                             0, false, storeOp.getNontemporal());
+
+        setTag(baseType, tree, storeOp.getContext(), op);
```

Comment:
Don't we need the same check (`if (isAnyOf<IndexType, IntegerType, FloatType, MemRefType>(baseType)) {`) here too?

Reply:
Good catch thanks!